cli/src/commands/deploy.ts

```diff
@@ -218,40 +225,43 @@ export default class Deploy extends Command {
       await remove(stagingDir);
     }
 
-    this.startStage('register app on local');
+    const updateAssociation = !givenAppName && !forceNewApp;
+    let application: Application;
     const env = 'default'; // hardcoded for now
-    let project = findProjectByDirectory(projects, projectDir);
+    const makeProject = (applicationId: string) => ({
+      directory: projectDir,
+      applicationId,
+      defaultEnv: env,
+    });
 
-    let application: Application;
-    if (!project || givenAppName) {
-      const applicationId = givenAppName !== undefined
-        ? await this.findApplicationIdByName(givenAppName)
-        : await this.selectApplicationForProject();
+    this.startStage('choose deployment target');
+    let project = forceNewApp ? undefined :
+      givenAppName ? makeProject(await this.findApplicationIdByName(givenAppName)) :
+      findProjectByDirectory(projects, projectDir);
 
+    if (!project && updateAssociation) {
+      const applicationId = await this.selectApplicationForProject();
       if (applicationId !== undefined) {
-        project = {
-          directory: projectDir,
-          applicationId,
-          defaultEnv: env,
-        };
-        // Don't update association if given --app-name option
-        if (!givenAppName) {
-          projects.push(project);
-          this.conf.set('projects', projects);
-        }
+        this.startStage('register app on local');
```

Comment:
1. This times local execution.
2. Having it conditional will make analytics harder.  Let's either not have a stage, or just `this.startStage('start')` before starting.